pkg/gpu/nvidia/mig/mig.go

```diff
@@ -0,0 +1,357 @@
+		devices[id] = pluginapi.Device{
+			ID:     id,
+			Health: pluginapi.Healthy,
+		}
+	}
+
+	return devices
+}
+
+// DeviceSpec returns the device spec that inclues list of devices to allocate for a deviceID.
+func (d *DeviceManager) DeviceSpec(deviceID string) ([]pluginapi.DeviceSpec, error) {
+	deviceSpecs, ok := d.gpuPartitions[deviceID]
+	if !ok {
+		return []pluginapi.DeviceSpec{}, fmt.Errorf("invalid allocation request with non-existing GPU partition: %s", deviceID)
+	}
+
+	return deviceSpecs, nil
+}
+
+// Start method performs the necessary initializations and starts the mig.DeviceManager.
+func (d *DeviceManager) Start() error {
+	if err := nvml.Init(); err != nil {
+		return fmt.Errorf("unable to initialize nvml: %v", err)
+	}
+
+	gpuCount, err := nvml.GetDeviceCount()
+	if err != nil {
+		return fmt.Errorf("failed to get device count: %v", err)
+	}
+
+	for i := uint(0); i < gpuCount; i++ {
+		device, err := nvml.NewDevice(i)
+		if err != nil {
+			return fmt.Errorf("failed to look up device index: %d, error: %v", i, err)
+		}
+
+		d.devices = append(d.devices, device)
+	}
+	return nil
+}
+
+// Stop method performs necessary cleanup operations to safely shut down mig.DeviceManager.
+func (d *DeviceManager) Stop() error {
+	return nvml.Shutdown()
+}
+
+// ConfigureGPUPartitions accepts a string describing the GPU partition type, and partitions all the GPUs on the node as specified.
+func (d *DeviceManager) ConfigureGPUPartitions(partition string) error {
+	if err := d.enableMIGMode(); err != nil {
+		return fmt.Errorf("failed to configure GPU partitions: %v", err)
+	}
+
+	if err := deleteAllGPUPartitions(); err != nil {
+		return fmt.Errorf("failed to delete existing GPU partitions: %v", err)
+	}
+
+	for _, device := range d.devices {
+		if err := createPartitionsOnGPU(partition, device); err != nil {
+			return fmt.Errorf("failed to create GPU partitions on device: %s, error: %v", device.Path, err)
+		}
+	}
+
+	if err := d.createDeviceNodes(); err != nil {
+		return fmt.Errorf("failed to create device node: %v", err)
+	}
+
+	return nil
+}
+
+// enableMIGMode ensures that MIG mode is enabled on all GPUs. If all GPUs are already in MIG mode, no action is taken.
+// If at least one GPU had to be flipped into MIG mode, the node is restarted as well.
+func (d *DeviceManager) enableMIGMode() error {
+	rebootNeeded := false
+
+	for _, device := range d.devices {
+		migEnabled, err := device.IsMigEnabled()
+		if err != nil {
+			return fmt.Errorf("failed to check if MIG is enabled on device %s: %v", device.Path, err)
+		}
+		if migEnabled {
+			continue
+		}
+
+		// If at least one GPU was not already in MIG mode, we need to reboot the VM.
+		rebootNeeded = true
+
+		_, err = device.SetMigMode(nvml.DEVICE_MIG_ENABLE)
+		if err != nil {
+			return fmt.Errorf("could not enable mig mode on GPU %s: %v", device.Path, err)
+		}
+	}
+
+	if rebootNeeded {
+		glog.Infof("Node needs to rebooted after turning on MIG mode on GPUs")
+		err := rebootNode()
```

Comment:
this doesnt seem to be blocking. So it is possible that `enableMIGMode()` returns with no error but the node is not actually in MIG mode? Should we return an error here telling the caller that the node is not yet in MIG mode?